### PR TITLE
add reports to detect existence of cocina property in dros and collections

### PR DESCRIPTION
**Pull request by ndushay**
## Why was this change made? 🤔

- for #4244, sul-dlss/cocina-models/issues/532 ...
- To create a general tool that can be used by devs or @arcadiafalcone for this purpose.

## How was this change tested? 🤨

I ran reports for sul-dlss/cocina-models/issues/532 and for #4244 with this, and found some data with #4244 ...

⚡ ⚠ If this change has cross service impact, including data writes to shared file systems, ***run [integration tests](https://github.com/sul-dlss/infrastructure-integration-test)*** and/or test in [stage|qa] environment, in addition to specs. ⚡

### CODE REVIEW

File: app/reports/property_existence_collections.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  # NOTE: Prefer strict JSON querying over lax when using the `.**` operator, per
+  #       https://www.postgresql.org/docs/14/functions-json.html#STRICT-AND-LAX-MODES
+  #
+  # > The .** accessor can lead to surprising results when using the lax mode.
+  # > ... This happens because the .** accessor selects both the segments array
+  # > and each of its elements, while the .HR accessor automatically unwraps
+  # > arrays when using the lax mode. To avoid surprising results, we recommend
+  # > using the .** accessor only in the strict mode.
+  JSONB_PATH = 'strict $.**.groupedValue ? (@.size() > 0)' # when property is array
+  # JSONB_PATH = 'strict $.**.contributor.type' # when property is a string - maybe keep size check to avoid empty values
```
Comment (mjgiarlo) [inline app/reports/property_existence_collections.rb:17]: cruft?

File: app/reports/property_existence_collections.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+class PropertyExistenceCollections
+  # NOTE: Prefer strict JSON querying over lax when using the `.**` operator, per
+  #       https://www.postgresql.org/docs/14/functions-json.html#STRICT-AND-LAX-MODES
+  #
+  # > The .** accessor can lead to surprising results when using the lax mode.
+  # > ... This happens because the .** accessor selects both the segments array
+  # > and each of its elements, while the .HR accessor automatically unwraps
+  # > arrays when using the lax mode. To avoid surprising results, we recommend
+  # > using the .** accessor only in the strict mode.
+  JSONB_PATH = 'strict $.**.groupedValue ? (@.size() > 0)' # when property is array
```
Comment (mjgiarlo) [inline app/reports/property_existence_collections.rb:16]: So... report on any non-empty grouped values?

File: app/reports/property_existence_dros.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  # NOTE: Prefer strict JSON querying over lax when using the `.**` operator, per
+  #       https://www.postgresql.org/docs/14/functions-json.html#STRICT-AND-LAX-MODES
+  #
+  # > The .** accessor can lead to surprising results when using the lax mode.
+  # > ... This happens because the .** accessor selects both the segments array
+  # > and each of its elements, while the .HR accessor automatically unwraps
+  # > arrays when using the lax mode. To avoid surprising results, we recommend
+  # > using the .** accessor only in the strict mode.
+  JSONB_PATH = 'strict $.**.contributor.parallelContributor ? (@.size() > 0)' # when property is array
+  # JSONB_PATH = 'strict $.**.contributor.type' # when property is a string - maybe keep size check to avoid empty values
```
Comment (mjgiarlo) [inline app/reports/property_existence_dros.rb:17]: Cruft?

File: app/reports/property_existence_dros.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+class PropertyExistenceDros
+  # NOTE: Prefer strict JSON querying over lax when using the `.**` operator, per
+  #       https://www.postgresql.org/docs/14/functions-json.html#STRICT-AND-LAX-MODES
+  #
+  # > The .** accessor can lead to surprising results when using the lax mode.
+  # > ... This happens because the .** accessor selects both the segments array
+  # > and each of its elements, while the .HR accessor automatically unwraps
+  # > arrays when using the lax mode. To avoid surprising results, we recommend
+  # > using the .** accessor only in the strict mode.
+  JSONB_PATH = 'strict $.**.contributor.parallelContributor ? (@.size() > 0)' # when property is array
```
Comment (mjgiarlo) [inline app/reports/property_existence_dros.rb:16]: I'm trying to square the names of the reports of this PR with these JSON paths. This seems narrower than "report on occurrences of a property." Intended or oversight or a misunderstanding on my part?

File: app/reports/property_existence_dros.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  def self.report
+    puts "item_druid,catkey,collection_druid,collection_name\n"
+    rows(SQL).compact.each { |row| puts row }
+  end
+
+  def self.rows(sql_query)
+    sql_result_rows = ActiveRecord::Base.connection.execute(sql_query).to_a
+
+    sql_result_rows.map do |row|
+      collection_druid = row['collection_druid']
```
Comment (mjgiarlo) [inline app/reports/property_existence_dros.rb:35]: This is `collection_id` above, no?

File: app/reports/property_existence_collections.rb (hunk 1 above)
Comment (ndushay) [reply app/reports/property_existence_collections.rb:17]: no - when property is an array vs when property is a string

File: app/reports/property_existence_collections.rb (hunk 2 above)
Comment (ndushay) [reply app/reports/property_existence_collections.rb:16]: > So... report on any non-empty grouped values?

yes - that's what this report is doing - reporting on non-empty occurrences of `groupedValue`

File: app/reports/property_existence_dros.rb (hunk 3 above)
Comment (ndushay) [reply app/reports/property_existence_dros.rb:17]: same as above:  one for array property, one for string.

File: app/reports/property_existence_dros.rb (hunk 4 above)
Comment (ndushay) [reply app/reports/property_existence_dros.rb:16]: I used a generic name for the report so Arcadia knows what the report does in the abstract so she will use it when she wants to do a run.

File: app/reports/property_existence_collections.rb (hunk 2 above)
Comment (mjgiarlo) [reply app/reports/property_existence_collections.rb:16]: Awesome.

File: app/reports/property_existence_dros.rb (hunk 4 above)
Comment (mjgiarlo) [reply app/reports/property_existence_dros.rb:16]: kk

**Review by mjgiarlo — APPROVED** [review]

**Pull request merged**